src/common/servers/serverManager.ts

```diff
@@ -0,0 +1,426 @@
+
+        return this.serverOrder.reduce((servers, srv) => {
+            const server = this.servers.get(srv);
+            if (server) {
+                servers.push(server);
+            }
+            return servers;
+        }, [] as MattermostServer[]);
+    }
+
+    getCurrentServer = () => {
+        log.debug('getCurrentServer');
+
+        if (!this.currentServerId) {
+            throw new Error('No server set as current');
+        }
+        const server = this.servers.get(this.currentServerId);
+        if (!server) {
+            throw new Error('Current server does not exist');
+        }
+        return server;
+    }
+
+    getLastActiveTabForServer = (serverId: string) => {
+        log.debug('ServerManager.getLastActiveTabForServer', serverId);
+
+        const lastActiveTab = this.lastActiveTab.get(serverId);
+        if (lastActiveTab) {
+            const tab = this.tabs.get(lastActiveTab);
+            if (tab && tab?.isOpen) {
+                return tab;
+            }
+        }
+        return this.getFirstOpenTabForServer(serverId);
+    }
+
+    getServer = (id: string) => {
+        return this.servers.get(id);
+    }
+
+    getTab = (id: string) => {
+        return this.tabs.get(id);
+    }
+
+    getAllServers = () => {
+        return [...this.servers.values()];
+    }
+
+    hasServers = () => {
+        return Boolean(this.servers.size);
+    }
+
+    getRemoteInfo = (serverId: string) => {
+        return this.remoteInfo.get(serverId);
+    }
+
+    updateRemoteInfos = (remoteInfos: Map<string, RemoteInfo>) => {
+        let hasUpdates = false;
+        remoteInfos.forEach((remoteInfo, serverId) => {
+            this.remoteInfo.set(serverId, remoteInfo);
+            hasUpdates = this.updateServerURL(serverId) || hasUpdates;
+            hasUpdates = this.openExtraTabs(serverId) || hasUpdates;
+        });
+
+        if (hasUpdates) {
+            this.persistServers();
+        }
+    }
+
+    lookupTabByURL = (inputURL: URL | string, ignoreScheme = false) => {
+        log.silly('lookupTabByURL', `${inputURL}`, ignoreScheme);
+
+        const parsedURL = urlUtils.parseURL(inputURL);
+        if (!parsedURL) {
+            return undefined;
+        }
+        const server = this.getAllServers().find((server) => {
+            return equalUrlsIgnoringSubpath(parsedURL, server.url, ignoreScheme) && parsedURL.pathname.match(new RegExp(`^${server.url.pathname}(.+)?(/(.+))?$`));
+        });
+        if (!server) {
+            return undefined;
+        }
+        const tabs = this.getOrderedTabsForServer(server.id);
+
+        let selectedTab = tabs.find((tab) => tab && tab.name === TAB_MESSAGING);
+        tabs.
+            filter((tab) => tab && tab.name !== TAB_MESSAGING).
+            forEach((tab) => {
+                if (parsedURL.pathname.match(new RegExp(`^${tab.url.pathname}(/(.+))?`))) {
+                    selectedTab = tab;
+                }
+            });
```

Comment:
Yeah this logic is very confusing, it's been around a while in various forms. What this is supposed to do is find the exact view that best matches an arbitrary URL that you provide.

The reason we iterate so many times is that we basically assume that the view of type `TAB_MESSAGING` (which has a url of the form `http://<somehost>`) always is the default case (and it should be, it's the Channels tab), so we grab that first as the default case. Then we check the other tabs, which currently have a base URL of the form `http://<somehost>/boards` or `http://<somehost>/playbooks` to see if either of them are better matches.

We do the `find()` first to separate out the Channels case because it will always match the above regex, and since we don't guarantee any order, there are cases where you click on a Boards link and it would forward you to the Channels tab because that's the last one selected.

A lot of this logic was shoehorned in when we added the tabs and it's complicated and annoying, and one of the biggest reasons I want to move to arbitrary tabs is to get rid of this. We could as a short term fix patch the regex to only look for `boards` or `playbooks`, but I don't see a huge benefit in potentially causing some issue now when I fully plan on getting rid of this logic altogether.

Reply:
is fine like this.. is not like this is a massive perf issue, but could definitely benefit from some logic change.